openraft/tests/life_cycle/t20_initialization.rs

```diff
@@ -229,6 +229,151 @@ async fn initialize_err_not_allowed() -> anyhow::Result<()> {
     Ok(())
 }
 
+/// What does this test do?
+///   this test is ought to test if the router is aware of network failures
+///
+/// - initialize 1 node as leader and 2 nodes as follower
+/// - set one follower's network to be unreachable
+/// - check if router's connecting to unreachable follower returning an `Err`
+/// - assert whether the cluster still works properly
+#[async_entry::test(worker_threads = 8, init = "init_default_ut_tracing()", tracing_span = "debug")]
+async fn router_network_failure_aware() -> anyhow::Result<()> {
+    let config = Arc::new(Config::default().validate()?);
+    let nodes = btreeset! {0, 1, 2};
+    let mut router = RaftRouter::new(config.clone());
+    let mut log_index = 0;
+
+    // making cluster by hand
+    tracing::info!("--- init single node");
+    {
+        router.new_raft_node(0);
+        router.initialize_from_single_node(0).await?;
+        log_index += 1;
+        router.wait_for_log(&btreeset! {0}, Some(log_index), timeout(), "init single node cluster").await?;
+    }
+
+    tracing::info!("--- add reachable learners to cluster");
+    {
+        router.new_raft_node(1);
+        router.add_learner(0, 1).await?;
+        log_index += 1;
+
+        router.new_raft_node(2);
+        router.add_learner(0, 2).await?;
+        log_index += 1;
+
+        router
+            .wait_for_log(
+                &btreeset! {0},
+                Some(log_index),
+                timeout(),
+                "should add 2 learner the cluster",
+            )
+            .await?;
+    }
+
+    tracing::info!("--- add unreachable learner to cluster");
+    {
+        router.new_raft_node(3);
+        router.block_node(3);
+        assert!(router.add_learner(0, 3).await.is_err());
+
+        router
+            .wait_for_log(
+                &btreeset! {0, 1, 2},
+                Some(log_index),
+                timeout(),
+                "no logs commit in the cluster",
+            )
+            .await?;
+    }
+
+    tracing::info!("change membership");
+    {
+        assert!(router.leader().is_some());
+
+        let leader = 0;
+        let h = router.get_raft_handle(&leader)?;
+
+        // add n3 should fail
+        assert!(h.change_membership(btreeset! {0, 1, 2, 3}, true, false).await.is_err());
```

Comment:
Why does this change-membership fail? There is an alive quorum IMO.

Reply:
> Why does this change-membership fail? There is an alive quorum IMO.

Because node 3 is not a learner in the cluster.

Reply:
Then why do you need to test this behavior?

Reply:
You are right, it does not worth it.